src/benchmarks/Sofa.Type/Quat.cpp

```diff
@@ -0,0 +1,180 @@
+#include <benchmark/benchmark.h>
+
+#include <utils/RandomValuePool.h>
+
+#include <sofa/type/Vec.h>
+#include <sofa/type/Quat.h>
+
+#include <array>
+#include <numeric>
+
+constexpr int64_t minSubIterations = 8 << 10;
+constexpr int64_t maxSubIterations = 8 << 10;
+
+static void BM_Quat_rotateVec(benchmark::State& state);
+BENCHMARK(BM_Quat_rotateVec)->RangeMultiplier(2)->Ranges({ {minSubIterations, maxSubIterations} })->Unit(benchmark::kMicrosecond);
+static void BM_Quat_rotateVec_impl_new(benchmark::State& state);
+BENCHMARK(BM_Quat_rotateVec_impl_new)->RangeMultiplier(2)->Ranges({ {minSubIterations, maxSubIterations} })->Unit(benchmark::kMicrosecond);
+static void BM_Quat_rotateVec_impl_old(benchmark::State& state);
+BENCHMARK(BM_Quat_rotateVec_impl_old)->RangeMultiplier(2)->Ranges({ {minSubIterations, maxSubIterations} })->Unit(benchmark::kMicrosecond);
+
+static void BM_Quat_invrotateVec(benchmark::State& state);
+BENCHMARK(BM_Quat_invrotateVec)->RangeMultiplier(2)->Ranges({ {minSubIterations, maxSubIterations} })->Unit(benchmark::kMicrosecond);
+
+void BM_Quat_rotateVec(benchmark::State& state)
+{
+    using Quat = sofa::type::Quatf;
+    using Vec3 = sofa::type::Vec3f;
+
+    constexpr auto totalsize = maxSubIterations * 3 + maxSubIterations * 4;
+    const auto& quatValues = RandomValuePool<float, totalsize>::get(0.f, 1.f);
+    const auto& vectValues = RandomValuePool<float, totalsize>::get();
+    std::vector<Quat> quat;
+    std::vector<Vec3> vect;
+
+    for (auto _ : state)
+    {
+        state.PauseTiming();
+        quat.resize(state.range(0));
+        vect.resize(state.range(0));
+
+        for (unsigned int i = 0; i < state.range(0); ++i)
+        {
+            quat.emplace_back(Quat{ quatValues[i*4 + 0], quatValues[i * 4 + 1], quatValues[i * 4 + 2], quatValues[i * 4 + 3] });
+            vect.emplace_back(Vec3{ vectValues[i*3 + 0], vectValues[i * 3 + 1], vectValues[i * 3 + 2] });
```

Comment:
```suggestion
            quat.emplace_back( quatValues[i*4 + 0], quatValues[i * 4 + 1], quatValues[i * 4 + 2], quatValues[i * 4 + 3] );
            vect.emplace_back( vectValues[i*3 + 0], vectValues[i * 3 + 1], vectValues[i * 3 + 2] );
```

The point of `emplace_back` is to avoid constructing the object by yourself. It is constructed with the parameters you give at the right location directly in the vector.

Another point: you resize and then emplace_back. Each emplace_back increases the size of the vector. You probably wanted to reserve, instead of resize.

Reply:
Exact! thanks! 👍
But i dont think it will change the timings, as the timer is paused at this location

Reply:
Yes, but you don't measure the same thing